File: Matlab Code/Classes/GrowthModel.m
classdef GrowthModel < handle
    
    % GrowthModel is an Abstract class which will be an interface for the
    % GrowthModels that are used in Imagine.
    % It will be responsible for modelling the growth of a crop, and so
    % will have it's state, a state propagation function and output
    % functions. 
    % The GrowthModel will also need to respond to Events (ImagineEvents,
    % not MATLAB object events. These events may or may not alter the state
    % and may or may not produce saleable products.
    % A GrowthModel will also need to inform other objects what 
    % ImagineEvents and Products are supported. So while the productNames
    % will be and imagineEventNames will be protected, we need to let other
    % objects get them.
    
    properties (SetAccess = protected)
        
        % An instance of a concrete subclass of the GrowthModelDelegate
        % class. The delegate actually contains all the information about
        % the growth model. This class merely acts as a standard wrapper so
        % that we can have arrays of GrowthModels, even while the models
        % may be very different.
        delegate
        
        % A string with the name of the class of the GrowthModel's delegate. 
        delegateClass
        
    end
    
    properties (Constant)
       transitionFunctionPrefix = 'transitionFunction_'; 
    end
    
    properties (Dependent)

        % These are dependent properties because they will end up coming
        % from the delegate. They have the same comments in the Abstract
        % GrowthModelDelegate class.
        
        % A String that gives the name of this growth model. It will be set
        % to the delegate's name, so is dependent.
        name
                
        % A list of ImagineEvents, which will come from the delegate.
        % It's comprised of the initial, regular and destruction events
        % that are stored in the growthModelDelegate.
        growthModelEvents
        growthModelInitialEvents
        growthModelRegularEvents
        growthModelDestructionEvents
        growthModelFinancialEvents
        
                
        % A list of strings with the names of ImagineEvents that have
        % transition functions defined in the delegate. That is, the
        % transitionFunction(eventName, ...) will have defined behaviour
        % for the given eventName.
        supportedImagineEvents
        
        % A list of strings with the names of the CropCategories that this
        % GrowthModel will be appropriate to model.
        supportedCategories
        
        % productPriceModels - a PriceModel for each product that can be
        % produced by this growthModel. The Units of the products are found
        % as the denominator of the PriceModels. 
        % When a product is produced, the Amount's unit can be used to work
        % out which price to use. It will match the denominator unit. (The
        % numerator will be in currency, probably $).
        productPriceModels
        
        % productUnits - matches the productPriceModels denominatorUnits.
        productUnits
        
        % growthModelOutputUnits - a list of Units that provide list of the
        % types of outputs the growthModel provides when passed the state
        % of a PlantedCrop. This list should include the union of all the
        % coreCropOutputUnits provided by the cropCategorys that this
        % growthModel supports.
        growthModelOutputUnits
        
        % growthModelOutputRates - a list of 0 rates that provide the
        % rate units that come out of the calculateOuputs function.
        % The rates can be important if you need to know how the
        % denominator will come out.
        % There is a case where the denominator will be nothing or just
        % 'unit' when it doesn't make sense for the rate to be a rate, but
        % an amount. Eg, height of Wheat. Doesn't make sense to divide by
        % Hectares. Or even more starkly - colour of wheat. This is a
        % weird but possibly useful output, but would be silly for it to be a rate.
        % So when a rate is just 500 somethings per 1, that's just 500
        % somethings. This will be appropriate if we want to make 'rainfall
        % that's fallen in the last 3 months' an output. It's not 'per'
        % anything. It just is.
        growthModelOutputRates
        
       
        % This will be a string cell array with descriptions of each
        % element of the state.
        % Comes from the delegate.
%?
        stateDescription
        
        % Number of elements in the state.
        % Comes from the delegate.
% ?
        stateSize

    end
    
    methods
        % Concrete subclasses of GrowthModelDelegate should register
        % themselves in this constructor as a simple way of creating a
        % GrowthModel with the appropriate delgate.
        function gm = GrowthModel(gmName)
            if nargin >= 1
               if ischar(gmName)
                    
                   switch gmName
                       
                       case 'Rainfall Based'
                           gm.delegate = RainfallBasedGrowthModelDelegate();
                           gm.delegateClass = class(gm.delegate);

                       case 'Annual'
                           gm.delegate = AnnualGrowthModelDelegate();
                           gm.delegateClass = class(gm.delegate);
                           
                       case 'AB-Gompertz'
                           gm.delegate = ABGompertzGrowthModelDelegate();
                           gm.delegateClass = class(gm.delegate);  
                           
                       case 'Fixed Yield'
                           gm.delegate = FixedYieldGrowthModelDelegate();
                           gm.delegateClass = class(gm.delegate);  
                       
                       case 'Simple Pasture'
                           gm.delegate = SimplePastureGrowthModelDelegate();
                           gm.delegateClass = class(gm.delegate);  
                           
                       case 'Imported GrassGro Data Model'
                           gm.delegate = GrassGroGrowthModelDelegate();
                           gm.delegateClass = class(gm.delegate);  

                       otherwise
                           error('GrowthModel name provided is not supported in constructor.');
                   end
                   
               end
            end         
        end
    end
    
    methods (Static)
        % checks that the growthModel is a ValidGrowthModel. It needn't
        % have valid parameters to be a valid object.
        function valid = isValid(gm)
            valid = isa(gm, 'GrowthModel') && ~isempty(gm);
            if valid && ~isempty(gm.delegate)
                valid = GrowthModelDelegate.isValid(gm.delegate);
            end
        end
        
        % checks that the parameters are ok. 
        function ready = isReady(gm)
            ready = GrowthModel.isValid(gm);
            if ready
               ready = GrowthModelDelegate.isReady(gm.delegate); 
            end
        end
        
        % Need to convert old RainfallBasedGrowthModels to
        % AnnualGrowthModels.
        function obj = loadobj(obj)

            if strcmp(obj.delegateClass, 'RainfallBasedGrowthModelDelegate')
                obj.delegateClass = 'AnnualGrowthModelDelegate';
                pPMs = obj.productPriceModels;
                gmEs = obj.growthModelEvents;
                obj.delegate = AnnualGrowthModelDelegate.convertFromRainfallBasedGMDelegate(obj.delegate);
                obj.productPriceModels = pPMs;
                obj.growthModelEvents = gmEs;
            end
        end
    end
    
    methods
       % These methods are simply the get and set methods for the delegate variables
       % that wrap to the main GrowthModel class.
       
       function name = get.name(gm)
           disp(gm.delegate)
           if ~isempty(gm.delegate)
               name = gm.delegate.modelName; 
               disp(name)
           else
               name = '';
           end
           disp(name)
       end       
       function sIEs = get.supportedImagineEvents(gm)
           if ~isempty(gm.delegate)
               sIEs = gm.delegate.supportedImagineEvents;
           else
               sIEs = {};
           end     
       end
       
       function sCs = get.supportedCategories(gm)        
           if ~isempty(gm.delegate)
               sCs = gm.delegate.supportedCategories;    
           else
               sCs = {};
           end              
       end
           
       function stSize = get.stateSize(gm)
           if ~isempty(gm.delegate)
               stSize = gm.delegate.stateSize;
           else
               stSize = 0;
           end
       end
       
       function stDes = get.stateDescription(gm)
           if ~isempty(gm.delegate)
               stDes = gm.delegate.stateDescription;
           else
               stDes = {};
           end
       end
       
       function gmEvts = get.growthModelInitialEvents(gm)
           if ~isempty(gm.delegate)
               gmEvts = gm.delegate.growthModelInitialEvents;
           else
               gmEvts = ImagineEvent.empty(1, 0);
           end
       end
       
       function gmEvts = get.growthModelRegularEvents(gm)
           if ~isempty(gm.delegate)
               gmEvts = gm.delegate.growthModelRegularEvents;
           else
               gmEvts = ImagineEvent.empty(1, 0);
           end
       end
       
       function gmEvts = get.growthModelDestructionEvents(gm)
           if ~isempty(gm.delegate)
               gmEvts = gm.delegate.growthModelDestructionEvents;
           else
               gmEvts = ImagineEvent.empty(1, 0);
           end
       end
       
       
       function gmEvts = get.growthModelFinancialEvents(gm)
           if ~isempty(gm.delegate)
               gmEvts = gm.delegate.growthModelFinancialEvents;
           else
               gmEvts = ImagineEvent.empty(1, 0);
           end
       end
       
       
       function gmEvts = get.growthModelEvents(gm)
           if ~isempty(gm.delegate)
               gmEvts = [gm.delegate.growthModelInitialEvents, ...
                         gm.delegate.growthModelRegularEvents, ...
                         gm.delegate.growthModelDestructionEvents, ...
                         gm.delegate.growthModelFinancialEvents];
           else
               gmEvts = ImagineEvent.empty(1, 0);
           end
       end
       
       function set.growthModelEvents(gm, evts)
                  
           % If we set the long growthModelEvents list, we need to find the
           % corresonding events in the sub lists and set them.
           for i = 1:length(evts)
              
               ix = find(strcmp(evts(i).name, {gm.delegate.growthModelInitialEvents.name}), 1, 'first');
               if ~isempty(ix)
                   gm.delegate.growthModelInitialEvents(ix) = evts(i);
               else
                   ix = find(strcmp(evts(i).name, {gm.delegate.growthModelRegularEvents.name}), 1, 'first');
                   if ~isempty(ix)
                        gm.delegate.growthModelRegularEvents(ix) = evts(i);
                   else
                        ix = find(strcmp(evts(i).name, {gm.delegate.growthModelDestructionEvents.name}), 1, 'first');
                        if ~isempty(ix)
                            gm.delegate.growthModelDestructionEvents(ix) = evts(i);
                        end
                   end
               end
               
           end
           
       end
           
       function pPMs = get.productPriceModels(gm)
          if ~isempty(gm.delegate)
              pPMs = gm.delegate.productPriceModels;
          else
              pPMs = PriceModel.empty(1, 0);
          end
       end
       
       function set.productPriceModels(gm, pPMs)
          if ~isempty(gm.delegate)
               gm.delegate.productPriceModels = pPMs;
          end
       end
       
       function pUs = get.productUnits(gm)
          pUs = [gm.productPriceModels.denominatorUnit]; 
       end

       function gMOUs = get.growthModelOutputUnits(gm)
          if ~isempty(gm.delegate)
              gMOUs = gm.delegate.growthModelOutputUnits;
          else
              gMOUs = Unit.empty(1, 0);
          end
       end
       
       function gMORs = get.growthModelOutputRates(gm)
          if ~isempty(gm.delegate)
              gMORs = gm.delegate.growthModelOutputRates;
          else
              gMORs = Rate.empty(1, 0);
          end
       end       
       
    end
    
    methods
           
       % These concrete methods simply wrap the methods of the delegate.
        
        function [newState, productRates] = propagateState(gm, plantedCrop, sim)
            % Propagates state over one month.
            [newState, productRates] = gm.delegate.propagateState(plantedCrop, sim);
        end
        
        function gm = setupGrowthModel(gm, cropName)
            % This should be implemented as a GUI that asks the user for input
            % that determines the growthModel's parameters.
            gm.delegate = gm.delegate.setupGrowthModel(cropName);
        end
        
        function renderGrowthModel(gm, ax)
            % This method should render a summary of the growth model on the
            % axes ax. This will be called within the cropWizard and should be
            % a fixed size.
            gm.delegate.renderGrowthModel(ax);
        end

        function [productAmounts, eventOutputs] = eventTransition(gm, eventName, plantedCrop, sim)
            % Wraps to the delegate's method that deals with an event.
            % transition functions in the delegate should be of the form 
            % transitionFunction_EVENT_NAME(gmdel, plantedCrop, sim) 
            % where eventName = 'EVENT NAME'
            % 
            % The transition functions update (transition) the state inside
            % plantedCrop and return a list of Amounts, which are the
            % Amounts of product harvested in the event.
            eventName = strrep(eventName, ' ', '_');
            [productAmounts, eventOutputs] = gm.delegate.([gm.transitionFunctionPrefix, eventName])(plantedCrop, sim);    
        end
        
        function outputsColumn = calculateOutputs(gm, state)
            % Wraps to the delegate's method that calculates the outputs
            % based on a PlantedCrop's state.
            outputsColumn = gm.delegate.calculateOutputs(state);
        end
        
        function cropNameHasChanged(gm, previousName, newName)
            gm.delegate.cropNameHasChanged(previousName, newName);
        end
    end
    
    methods
       
        function [productRates, outputRates] = getProductAndOutputRatesForEvent(gm, eventName)
            productRates = Rate.empty(1, 0);
            outputRates = Rate.empty(1, 0);
            try
                [productRates, outputRates] = gm.delegate.([gm.transitionFunctionPrefix, underscore(eventName)])([], []);
            catch ME %#ok<NASGU>
                return
            end
        end
        
        function productRates = getPropagationProductRates(gm)
            [~, productRates] = gm.delegate.propagateState([], []);
        end
    end
          
end

File: imagine/crop/growth_model.py
import abc
from imagine.util.underscore import underscore
from dotwiz import DotWiz

""" In Matlab Imagine, the GrowthModel class wrapped an Abstract GrowthModelDelegate class.
    This was due to the need to have an array of GrowthModels of the same type. Each growth model object
    would have a delegate (with the delegate being a specific concrete instance that subclassed an
    abstract GrowthModelDelegate class). The GrowthModel class simply wrapped the delegate class.
"""

"""
    % GrowthModel is an abstract class which will be an interface for the
    % GrowthModels that are used in Imagine.
    % It will be responsible for modelling the growth of a crop, and so
    % will have it's state, a state propagation function and output
    % functions. 
    % The GrowthModel will also need to respond to Events (ImagineEvents,
    % not MATLAB object events. These events may or may not alter the state
    % and may or may not produce saleable products.
    % A GrowthModel will also need to inform other objects what 
    % ImagineEvents and Products are supported. So while the productNames
    % will be and imagineEventNames will be protected, we need to let other
    % objects get them.
    
"""

"""
    The growth model doesn't retain state. It takes state as input, performs modifications and returns state as output.
    State is retained inside a PlantedCrop.
"""



class GrowthModel(abc.ABC):

    transition_function_prefix = "transition_function_"

    def __init__(self):
        self._supported_imagine_events = None
        self._growth_model_initial_events = []
        self._growth_model_regular_events = []
        self._growth_model_destruction_events = []
        self._growth_model_financial_events = []

    @abc.abstractmethod
    def load_gm(self, d):
        pass

    @property
    def growth_model_initial_events(self):
        return self._growth_model_initial_events

    @property
    def growth_model_regular_events(self):
        return self._growth_model_regular_events

    @property
    def growth_model_destruction_events(self):
        return self._growth_model_destruction_events

    @property
    def growth_model_financial_events(self):
        return self._growth_model_financial_events


    @property
    # TODO: revise this to return a dict?
    def growth_model_events(self):
        evts = self.growth_model_initial_events + \
               self.growth_model_regular_events + \
               self.growth_model_destruction_events + \
               self.growth_model_financial_events
        return evts

    # TODO: remove this note once fully resolved.
    # Removing product_price_models from GrowthModel. The price models will be part of a Crop now.
    # May need to provide the product rates though. Perhaps that's done through empty transition function calls.
    # @property
    # @abc.abstractmethod
    # def product_price_models(self):
    #     pass

    # @property
    # @abc.abstractmethod
    # def product_price_rates(self):
    #     pass
    #
    # @property
    # @abc.abstractmethod
    # def cost_price_rates(self):
    #     pass

    # @property
    # @abc.abstractmethod
    # def growth_model_output_units(self):
    #     pass
    #
    # @property
    # @abc.abstractmethod
    # def growth_model_output_rates(self):
    #     pass

    @property
    @abc.abstractmethod
    def model_name(self):
        pass

    @property
    @abc.abstractmethod
    def supported_categories(self):
        pass

    @property
    @abc.abstractmethod
    def state_description(self):
        pass

    @abc.abstractmethod
    def gm_is_valid(self):
        pass

    @abc.abstractmethod
    def gm_is_ready(self):
        pass

    @abc.abstractmethod
    def propagate_state(self, planted_crop, sim):
        pass

    # @abc.abstractmethod
    # def setup_growth_model(self, crop_name):
    #     pass
    #
    # @abc.abstractmethod
    # def render_growth_model(self, ax):
    #     pass

    def event_transition(self, event_name, planted_crop, sim):
        func_name = self.transition_function_prefix + underscore(event_name)
        transition_func = getattr(self, func_name, None)
        if transition_func:
            return transition_func(planted_crop, sim)
        return [], []

    @abc.abstractmethod
    def calculate_outputs(self, state):
        pass

    # @abc.abstractmethod
    # def crop_name_has_changed(self, previous_name, new_name):
    #     pass

    @property
    def supported_imagine_events(self):
        if self._supported_imagine_events is None:
            trans_fcns = []
            for method_name in dir(self.__class__):
                if method_name.startswith(self.transition_function_prefix):
                    fcn = method_name[len(self.transition_function_prefix):].replace('_', ' ')
                    trans_fcns.append(fcn)
            return trans_fcns
        else:
            return self._supported_imagine_events

    @property
    def state_size(self):
        return len(self.state_description)

    # TODO: remove these methods. Use the get___rates methods below instead.
    # def get_event_output_units(self, event_name):
    #     method_name = self.transition_function_prefix + event_name.replace(' ', '_')
    #     _, eos = getattr(self, method_name)(None, None)
    #     event_output_units = [eo.unit for eo in eos]
    #     return event_output_units
    #
    # def get_output_product_units(self, event_name):
    #     method_name = self.transition_function_prefix + event_name.replace(' ', '_')
    #     ops, _ = getattr(self, method_name)(None, None)
    #     event_output_units = [op.unit for op in ops]
    #     return event_output_units

    # These 3 methods get columns of empty Rates that will have the units and denominator units
    # that will be produced by the growth model.
    #
    #
    # Get product and event output rates for the given event.
    def get_event_output_rates(self, event_name):
        method_name = self.transition_function_prefix + event_name.lower().replace(' ', '_')
        if not hasattr(self, method_name):
            return [], []
        product_rates, event_output_rates = getattr(self, method_name)(None, None)
        return product_rates, event_output_rates

    # Get the output rates. (I.e. from calculate_outputs which is the view of the state.)
    def get_output_rates(self):
        output_rates = self.calculate_outputs()
        return output_rates

    # Get the product rates from propagate state.
    def get_propagation_product_rates(self):
        _, product_rates = self.propagate_state(None, None)
        return product_rates

    # Get an event by its name,
    def get_event_by_name(self, event_name):
        found_evt = [evt for evt in self.growth_model_events if evt.name == event_name]
        if len(found_evt) == 1:
            return found_evt[0]
        else:
            raise ValueError(f"event_name {event_name} does not match exactly one of the growth model's events.")

# A helper to set up a structure to store unit lists.
def make_growth_model_units():
    out = DotWiz()
    out.outputs = {}
    out.propagation_products = {}
    out.products = {}
    out.event_outputs = {}

    # These default to empty lists. Only need to implement non-empty ones.
    out.outputs.numerator_units = []
    out.outputs.denominator_units = []
    out.propagation_products.numerator_units = []
    out.propagation_products.denominator_units = []

    # Index these by event name.
    out.products.numerator_units = {}
    out.products.denominator_units = {}
    out.event_outputs.numerator_units = {}
    out.event_outputs.denominator_units = {}

    return out
